### ISSUE TITLE
New Rule: no-promise-executor-throw

### ISSUE THREAD

**Issue by std4453** [eslint/eslint#17121]
### Rule details

Detect `throw`s in side new Promise() executor

### Related ECMAScript feature

Promise.constructor

### What type of rule is this?

Warns about a potential problem

### Example code

```js
new Promise((resolve, reject) => {
  throw new Error();
})
```

### Why should this rule be in the core instead of a plugin?

The [`no-promise-executor-return`](https://eslint.org/docs/latest/rules/no-promise-executor-return) rule warns about using the keyword `return` inside `new Promise()`, reasoning that:

> The return value of the executor is ignored. Returning a value from an executor function is a possible error because the returned value cannot be used and it doesn’t affect the promise in any way.

Meanwhile, `throw`-ing inside of `new Promise()` will result in an unhandled Promise rejection, which cannot be caught - both with `.catch()` and `catch` blocks around `new Promise()`. This will cause Node.js progress to terminate if [`--unhandled-rejection=throw`](https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode) (default since v15).

That's rarely the true intention of the code, and the author probably meant to `reject()` with said error.

I think this should be in core instead of as a plugin since `no-promise-executor-return` was already there, and this rule does merely the same thing, just with a different keyword. I'm not sure whether it would be better to add a parameter to `no-promise-executor-return`, however the name of the rule might block such changes.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

**Comment by mdjermanovic** [comment]
Hi @std4453, thanks for the issue!

> Meanwhile, `throw`-ing inside of `new Promise()` will result in an unhandled Promise rejection, which cannot be caught - both with `.catch()` and `catch` blocks around `new Promise()`.

It doesn't seem so:

> ```js
> new Promise((resolve, reject) => {
>   throw new Error("foo");
> }).catch(error => { console.log(error.message); }); // logs "foo"
> ```

**Comment by std4453** [comment]
@mdjermanovic I was wrong, sorry for your inconvenience 😢
